include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+/// the nettest will emit to overriden methods written by you. The default
+/// behavior of the router is to ignore all events.
+///
+/// ```
+/// class MyRouter : public mk::nettest::routers::Router {
+///  public:
+///   using mk::nettest::routers::Router::Router;
+///
+///   void on_log(const mk::nettest::event::LogEvent &event) override {
+///     // Your event handling code here. Remember that this is called
+///     // in the context of the FFI API's background thread.
+///     //
+///     // In this example we emit log messages on the standard error.
+///     std::clog << event.log_level << ": " << event.message.
+///   }
+///
+///  private:
+///   // Your implementation variables here
+/// }
+/// ```
+///
+/// ### 4. Instantiate your router specialization
+///
+/// Create an instance of your router that will be used by this nettest. You
+/// will later pass a pointer to this instance to methods emitting events.
+///
+/// ```
+/// MyRouter my_router;
+/// ```
+///
+/// ### 5. Start the Nettest
+///
+/// Start the nettest using the specific start method for the nettest
+/// that you want to run. After this nettest specific method is called,
+/// all the other methods are nettest-agnostic, so you can have a lot
+/// of common nettest-processing code. Make sure you check the value
+/// returned by the nettest-specific start method. In case of error log
+/// messages will be emitted through the router.
+///
+/// ```
+/// if (!nettest.start_whatsapp(&my_router)) {
+///   // TODO: your code for handling this failure here.
+///   return;
+/// }
+/// ```
+///
+/// ### 6. Dispatch nettest events
+///
+/// Dispatch nettest events. This is nettest-agnostic code. Calling the
+/// Nettest::route_next_event() method will cause any pending events
+/// to be extracted from the events queue and emitted through the router.
+///
+/// ```
+/// while (!nettest.is_done()) {
+///   nettest.route_next_event(&my_router);
+/// }
+/// ```
+///
+/// ### 7. Final remarks
+///
+/// If the nettest goes out of scope, it will not be interrupted, rather it
+/// will run until completion without routing events. To interrupt a nettest,
+/// use Nettest::interrupt(). When you need to share a Nettest instance
+/// between different threads for the purpose of interrupting a test, your
+/// best option is to use `new` or `std::make_shared`.
+///
+/// \see https://github.com/measurement-kit/measurement-kit/tree/master/example/nettest for usage examples.
+
+#include <assert.h>
+#include <stdint.h>
+
+#include <deque>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/ffi.h>
+
+/// Measurement Kit namespace.
+namespace mk {
+
+/// Namespace containing the nettest API.
+namespace nettest {
+
+/// Contains events definitions.
+namespace events {
```

Comment:
This namespace contains mechanically generated classes using `jinja2`. The documentation of the classes is not very informative, to be honest, as I mentioned above. One possible alternative would be to also include the documentation strings in the information passed to the `jinja2` template; too much complexity?